server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraDumbBlobStoreCache.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.james.blob.api.BlobId;
+import org.reactivestreams.Publisher;
+
+import com.datastax.driver.core.ConsistencyLevel;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Session;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class CassandraDumbBlobStoreCache implements DumbBlobStoreCache {
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectStatement;
+    private final PreparedStatement deleteStatement;
+    private final CassandraCacheConfiguration cacheConfiguration;
+
+    @Inject
+    public CassandraDumbBlobStoreCache(Session session, CassandraCacheConfiguration cacheConfiguration) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.insertStatement = prepareInsert(session);
+        this.selectStatement = prepareSelect(session);
+        this.deleteStatement = prepareDelete(session);
+        this.cacheConfiguration = cacheConfiguration;
+    }
+
+    @Override
+    public Publisher<Void> cache(BlobId blobId, byte[] bytes) {
+        Preconditions.checkNotNull(bytes, "Data must not be null");
+
+        return Mono.just(bytes)
+            .flatMap(data -> {
+                if (data.length > cacheConfiguration.getSizeThresholdInBytes()) {
+                    return Mono.empty();
+                }
+                return Mono.from(save(blobId, ByteBuffer.wrap(data, 0, data.length)));
+            });
+    }
+
+    private Publisher<Void> save(BlobId blobId, ByteBuffer data) {
+        return cassandraAsyncExecutor.executeVoid(
+            insertStatement.bind()
+                .setString(ID, blobId.asString())
+                .setBytes(DATA, data)
+                .setInt(TTL, Math.toIntExact(cacheConfiguration.getTtl().getSeconds()))
+                .setConsistencyLevel(ConsistencyLevel.ONE));
+    }
+
+    @Override
+    public Publisher<Void> cache(BlobId blobId, InputStream inputStream) {
+        Preconditions.checkNotNull(inputStream);
+
+        return Mono.using(() -> IOUtils.toByteArray(new BufferedInputStream(inputStream,
+                cacheConfiguration.getSizeThresholdInBytes() + 1)),
+            bytes -> Mono.from(cache(blobId, bytes)),
+            any -> closeInputStreamQuite(inputStream));
+    }
+
+    @Override
+    public Publisher<byte[]> read(BlobId blobId) {
+        return cassandraAsyncExecutor
+            .executeSingleRow(
+                selectStatement.bind()
+                    .setString(ID, blobId.asString())
+                    .setConsistencyLevel(ConsistencyLevel.ONE)
+                    .setReadTimeoutMillis(Math.toIntExact(cacheConfiguration.getTimeOut().toMillis()))
+            )
+            .onErrorResume(Mono::error)
+            .map(row -> row.getBytes(DATA).array());
```

Comment:
do you have any cassandra driver documentation stating you can do that? Retrieving the backing array of a ByteBuffer doesn't work in the general case

Reply:
I got your point here.

Reply:
Yes we have to copy the underlying array.

```
ByteBuffer byteBuffer = row.getBytes(DATA);
byte[] data = ByteBuffer.allocate(byteFuffer.remaining());
byteBuffer.put(data);
return data;
```

Should do the trick.